src/sidebar/components/test/annotation-editor-test.js

```diff
@@ -0,0 +1,313 @@
+      });
+
+      const call = fakeStore.createDraft.getCall(0);
+      assert.calledOnce(fakeStore.createDraft);
+      assert.equal(call.args[1].text, 'updated text');
+    });
+
+    describe('tag editing', () => {
+      it('adds tag when add callback invoked', () => {
+        const wrapper = createComponent();
+        wrapper.find('TagEditor').props().onAddTag('newTag');
+
+        const storeCall = fakeTagsService.store.getCall(0);
+        const draftCall = fakeStore.createDraft.getCall(0);
+
+        assert.deepEqual(storeCall.args[0], [{ text: 'newTag' }]);
+        assert.deepEqual(draftCall.args[1].tags, ['newTag']);
+      });
+
+      it('trims tag when add callback invoked', () => {
+        const wrapper = createComponent();
+        wrapper.find('TagEditor').props().onAddTag(' newTag ');
+
+        const storeCall = fakeTagsService.store.getCall(0);
+        const draftCall = fakeStore.createDraft.getCall(0);
+
+        assert.deepEqual(storeCall.args[0], [{ text: 'newTag' }]);
+        assert.deepEqual(draftCall.args[1].tags, ['newTag']);
+      });
+
+      it('does not add empty tags', () => {
+        const wrapper = createComponent();
+        wrapper.find('TagEditor').props().onAddTag('  ');
+
+        assert.notCalled(fakeTagsService.store);
+        assert.notCalled(fakeStore.createDraft);
+      });
+
+      it('does not add duplicate tags', () => {
+        const draft = fixtures.defaultDraft();
+        draft.tags = ['newTag'];
+        fakeStore.getDraft.returns(draft);
+
+        const wrapper = createComponent();
+        wrapper.find('TagEditor').props().onAddTag('newTag');
+
+        assert.equal(fakeTagsService.store.callCount, 0);
+        assert.equal(fakeStore.createDraft.callCount, 0);
+      });
+
+      it('removes tag when remove callback invoked', () => {
+        const draft = fixtures.defaultDraft();
+        draft.tags = ['newTag'];
+        fakeStore.getDraft.returns(draft);
+
+        const wrapper = createComponent();
+        wrapper.find('TagEditor').props().onRemoveTag('newTag');
+
+        const draftCall = fakeStore.createDraft.getCall(0);
+
+        assert.equal(fakeTagsService.store.callCount, 0);
+        assert.deepEqual(draftCall.args[1].tags, []);
+      });
+
+      it('does not remove non-existent tags', () => {
+        const draft = fixtures.defaultDraft();
+        fakeStore.getDraft.returns(draft);
+
+        const wrapper = createComponent();
+        wrapper.find('TagEditor').props().onRemoveTag('newTag');
+
+        assert.equal(fakeTagsService.store.callCount, 0);
+        assert.equal(fakeStore.createDraft.callCount, 0);
+      });
+    });
+
+    describe('saving the annotation', () => {
+      it('saves the annotation when save callback invoked', async () => {
+        const annotation = fixtures.defaultAnnotation();
+        const wrapper = createComponent({ annotation });
+
+        await wrapper.find('AnnotationPublishControl').props().onSave();
```

Comment:
This `await` line currently works fine because the `onSave` callback doesn't trigger any state changes, but if it did, you'd need to wrap it with `act` - pardon the async-keyword soup:

```js
await act(async () => {
  await wrapper.find(...).props().onSave();
});
```